File: scripts/check-reservations.js
#!/usr/bin/env node
/**
 * Gastropub Reservation Monitor
 * Checks OpenTable + Resy for available slots at Jake's shortlisted spots.
 * Dates: 2026-02-28, 2026-03-01, 2026-03-02, 2026-03-04, 2026-03-06
 * Party size: 2
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const STATE_FILE = path.join(__dirname, 'reservation-state.json');

const RESTAURANTS = [
  {
    name: 'The Marksman',
    area: 'Shoreditch',
    platform: 'opentable',
    url: 'https://www.opentable.co.uk/r/the-marksman-hackney-london',
  },
  {
    name: 'The Marksman',
    area: 'Shoreditch',
    platform: 'resy',
    slug: 'marksman-public-house',
  },
  {
    name: 'The Princess of Shoreditch',
    area: 'Shoreditch',
    platform: 'opentable',
    url: 'https://www.opentable.co.uk/the-princess-of-shoreditch',
  },
  {
    name: 'The Royal Oak Marylebone',
    area: 'Marylebone',
    platform: 'resy',
    slug: 'the-royal-oak-marylebone',
  },
];

const DATES = [
  '2026-02-28',
  '2026-03-01',
  '2026-03-02',
  '2026-03-04',
  '2026-03-06',
];

const PARTY_SIZE = 2;
const PREFERRED_TIME = '19:00'; // 7pm default

function loadState() {
  try {
    return JSON.parse(fs.readFileSync(STATE_FILE, 'utf8'));
  } catch {
    return { found: {}, lastRun: null };
  }
}

function saveState(state) {
  fs.writeFileSync(STATE_FILE, JSON.stringify(state, null, 2));
}

function slotKey(restaurant, date, times) {
  return `${restaurant.name}|${restaurant.platform}|${date}|${times.sort().join(',')}`;
}

async function checkOpenTable(page, restaurant, date) {
  const dateTime = `${date}T${PREFERRED_TIME}`;
  const url = `${restaurant.url}?covers=${PARTY_SIZE}&dateTime=${dateTime}`;
  
  try {
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 });
    // Extra wait for JS to render
    await page.waitForTimeout(5000);
    
    // Wait for time slots to appear
    await page.waitForSelector('[data-test="time-slot"], button[class*="timeslot"], button[class*="time-slot"], [class*="TimeSlot"], a[class*="slot"]', {
      timeout: 15000
    }).catch(() => null);

    // Extract available times
    const slots = await page.evaluate(() => {
      const selectors = [
        '[data-test="time-slot"]',
        'button[class*="timeslot"]',
        'button[class*="TimeSlot"]',
        '[class*="timeslot-button"]',
        'a[class*="time-slot"]',
      ];
      
      for (const sel of selectors) {
        const els = document.querySelectorAll(sel);
        if (els.length > 0) {
          return Array.from(els)
            .filter(el => !el.disabled && !el.classList.toString().includes('unavailable'))
            .map(el => el.textContent.trim())
            .filter(t => t.length > 0);
        }
      }
      
      // Fallback: look for any text that looks like a time
      const allText = document.body.innerText;
      const timeMatches = allText.match(/\b\d{1,2}:\d{2}\s*(AM|PM|am|pm)?\b/g);
      return timeMatches ? [...new Set(timeMatches)] : [];
    });

    return slots;
  } catch (err) {
    console.error(`  [OpenTable] Error checking ${restaurant.name} on ${date}: ${err.message}`);
    return [];
  }
}

async function checkResy(page, restaurant, date) {
  const url = `https://resy.com/cities/lon/${restaurant.slug}?date=${date}&seats=${PARTY_SIZE}`;
  
  try {
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 });
    await page.waitForTimeout(5000);
    
    // Wait for time slots
    await page.waitForSelector('[class*="ReservationButton"], [class*="time-slot"], button[class*="slot"], [data-test*="slot"]', {
      timeout: 15000
    }).catch(() => null);

    const slots = await page.evaluate(() => {
      const selectors = [
        '[class*="ReservationButton"]',
        'button[class*="slot"]',
        '[data-test*="time-slot"]',
        '[class*="timeslot"]',
        'li[class*="time"] button',
      ];
      
      for (const sel of selectors) {
        const els = document.querySelectorAll(sel);
        if (els.length > 0) {
          return Array.from(els)
            .filter(el => !el.disabled)
            .map(el => el.textContent.trim())
            .filter(t => /\d{1,2}:\d{2}/.test(t));
        }
      }
      return [];
    });

    return slots;
  } catch (err) {
    console.error(`  [Resy] Error checking ${restaurant.name} on ${date}: ${err.message}`);
    return [];
  }
}

async function main() {
  const state = loadState();
  const newFindings = [];

  console.log(`\n🍺 Gastropub Reservation Check — ${new Date().toISOString()}`);
  console.log(`Checking ${RESTAURANTS.length} restaurants × ${DATES.length} dates...\n`);

  const browser = await chromium.launch({
    headless: true,
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-http2',
      '--disable-blink-features=AutomationControlled',
      '--disable-dev-shm-usage',
    ],
  });

  try {
    const context = await browser.newContext({
      userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36',
      locale: 'en-GB',
      timezoneId: 'Europe/London',
      viewport: { width: 1280, height: 800 },
      extraHTTPHeaders: {
        'Accept-Language': 'en-GB,en;q=0.9',
      },
    });

    // Hide webdriver property
    await context.addInitScript(() => {
      Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
    });

    const page = await context.newPage();

    for (const restaurant of RESTAURANTS) {
      for (const date of DATES) {
        process.stdout.write(`  Checking ${restaurant.name} (${restaurant.platform}) on ${date}... `);

        let slots = [];
        if (restaurant.platform === 'opentable') {
          slots = await checkOpenTable(page, restaurant, date);
        } else if (restaurant.platform === 'resy') {
          slots = await checkResy(page, restaurant, date);
        }

        if (slots.length > 0) {
          const key = `${restaurant.name}|${restaurant.platform}|${date}`;
          const existing = state.found[key] || [];
          const newSlots = slots.filter(s => !existing.includes(s));

          if (newSlots.length > 0 || existing.length === 0) {
            newFindings.push({
              restaurant: restaurant.name,
              area: restaurant.area,
              platform: restaurant.platform,
              date,
              slots,
              isNew: newSlots.length > 0,
            });
            state.found[key] = slots;
          }

          console.log(`✅ ${slots.length} slot(s): ${slots.join(', ')}`);
        } else {
          console.log('❌ None');
        }

        // Small delay to be polite
        await page.waitForTimeout(2000);
      }
    }
  } finally {
    await browser.close();
  }

  state.lastRun = new Date().toISOString();
  saveState(state);

  if (newFindings.length > 0) {
    console.log('\n🎉 NEW AVAILABILITY FOUND:\n');
    for (const f of newFindings) {
      const dateLabel = new Date(f.date).toLocaleDateString('en-GB', { weekday: 'long', day: 'numeric', month: 'long' });
      console.log(`  🍽️  ${f.restaurant} (${f.area}) — ${dateLabel}`);
      console.log(`     Platform: ${f.platform}`);
      console.log(`     Times: ${f.slots.join(', ')}`);
      console.log();
    }
    // Output summary for cron notification
    process.stdout.write('\nNOTIFY:');
    const summary = newFindings.map(f => {
      const dateLabel = new Date(f.date).toLocaleDateString('en-GB', { weekday: 'short', day: 'numeric', month: 'short' });
      return `${f.restaurant} on ${dateLabel} (${f.slots.join('/')}) via ${f.platform}`;
    }).join('; ');
    console.log(summary);
  } else {
    console.log('\nNo new availability found.');
    console.log('NOTIFY:none');
  }
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: scripts/check-reservations.py
#!/usr/bin/env python3
"""
Gastropub Reservation Monitor
Checks Resy for available slots at Jake's shortlisted spots.
Uses Resy's internal API directly.
"""

import json
import os
import sys
import urllib.request
import urllib.parse
import urllib.error
from datetime import datetime

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
STATE_FILE = os.path.join(SCRIPT_DIR, "reservation-state.json")

RESY_API_KEY = "VbWk7s3L4KiK5fzlO7JD3Q5EYolJI7n5"
RESY_HEADERS = {
    "Authorization": f'ResyAPI api_key="{RESY_API_KEY}"',
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
    "Origin": "https://resy.com",
    "Referer": "https://resy.com/",
    "Content-Type": "application/json",
    "Accept": "application/json, text/plain, */*",
    "Cache-Control": "no-cache",
}

# London coordinates
LONDON_LAT = 51.5074
LONDON_LNG = -0.1278

RESTAURANTS = [
    {
        "name": "The Marksman",
        "area": "Shoreditch",
        "platform": "resy",
        "search_query": "Marksman Public House Shoreditch",
    },
    {
        "name": "The Royal Oak Marylebone",
        "area": "Marylebone",
        "platform": "resy",
        "search_query": "Royal Oak Marylebone",
    },
    {
        "name": "The Princess of Shoreditch",
        "area": "Shoreditch",
        "platform": "opentable",
        "opentable_url": "https://www.opentable.co.uk/the-princess-of-shoreditch",
        "opentable_rid": None,  # will try to find dynamically
    },
]

DATES = [
    "2026-02-28",
    "2026-03-01",
    "2026-03-02",
    "2026-03-04",
    "2026-03-06",
]

PARTY_SIZE = 2


def resy_request(method, path, params=None, body=None):
    """Make a request to Resy's API."""
    url = f"https://api.resy.com{path}"
    if params:
        url += "?" + urllib.parse.urlencode(params)

    data = json.dumps(body).encode() if body else None
    req = urllib.request.Request(url, data=data, headers=RESY_HEADERS, method=method)

    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        try:
            return json.loads(body)
        except Exception:
            return {"error": str(e), "body": body[:200]}
    except Exception as e:
        return {"error": str(e)}


def search_resy_venue(query):
    """Search Resy for a venue by name, return (venue_id, slug) or None."""
    result = resy_request(
        "POST",
        "/3/venuesearch/search",
        body={
            "availability": True,
            "page": 1,
            "per_page": 10,
            "types": ["venue"],
            "geo": {
                "latitude": LONDON_LAT,
                "longitude": LONDON_LNG,
                "radius": 10000,
            },
            "query": query,
            "slot_filter": {"day": DATES[0], "party_size": PARTY_SIZE},
        },
    )

    if "search" in result and "hits" in result.get("search", {}):
        hits = result["search"]["hits"]
        if hits:
            hit = hits[0]
            venue_id = hit.get("venue", {}).get("id", {}).get("resy") or hit.get("id")
            url_slug = hit.get("url_slug") or hit.get("slug")
            name = hit.get("name", "Unknown")
            return {"venue_id": venue_id, "url_slug": url_slug, "name": name, "raw": hit}
    return None


def get_resy_availability(venue_id, date):
    """Get available time slots for a venue on a given date."""
    result = resy_request(
        "GET",
        "/4/find",
        params={
            "lat": "0",
            "long": "0",
            "day": date,
            "party_size": str(PARTY_SIZE),
            "venue_id": str(venue_id),
        },
    )

    slots = []
    if "results" in result:
        for venue in result["results"].get("venues", []):
            for slot in venue.get("slots", []):
                time_str = slot.get("date", {}).get("start", "")
                if time_str:
                    # Extract just the time part
                    try:
                        t = datetime.fromisoformat(time_str.replace("Z", "+00:00"))
                        slots.append(t.strftime("%H:%M"))
                    except Exception:
                        slots.append(time_str[-8:-3])  # fallback slice
    return slots


def check_opentable_availability(restaurant, date):
    """
    Try to check OpenTable availability via their widget API.
    Returns list of available time strings.
    """
    # OpenTable is heavily JS-rendered; this is best-effort via their
    # widget loader which is lighter weight
    # We'll try to get the restaurant ID first via their suggest API
    try:
        query = urllib.parse.quote(restaurant["name"])
        url = f"https://www.opentable.co.uk/restaurant/profile/search?q={query}&lang=en-GB&covers={PARTY_SIZE}&dateTime={date}T19%3A00"
        req = urllib.request.Request(
            url,
            headers={
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36",
                "Accept": "application/json",
                "Referer": "https://www.opentable.co.uk/",
            },
        )
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read())
            print(f"  [OpenTable debug] Response: {str(data)[:200]}")
    except Exception as e:
        print(f"  [OpenTable] Could not reach API: {e}")
    return []  # placeholder — OpenTable needs browser automation


def load_state():
    try:
        with open(STATE_FILE) as f:
            return json.load(f)
    except Exception:
        return {"found": {}, "venue_ids": {}, "lastRun": None}


def save_state(state):
    with open(STATE_FILE, "w") as f:
        json.dump(state, f, indent=2)


def main():
    state = load_state()
    new_findings = []

    print(f"\n🍺 Gastropub Reservation Check — {datetime.utcnow().isoformat()}")
    print(f"Checking {len(RESTAURANTS)} restaurants × {len(DATES)} dates...\n")

    for restaurant in RESTAURANTS:
        name = restaurant["name"]
        platform = restaurant["platform"]

        if platform == "resy":
            # Look up venue ID (cache it)
            if name not in state.get("venue_ids", {}):
                print(f"  🔍 Searching Resy for '{name}'...")
                venue_info = search_resy_venue(restaurant["search_query"])
                if venue_info:
                    print(f"     Found: {venue_info['name']} (ID: {venue_info['venue_id']}, slug: {venue_info['url_slug']})")
                    state.setdefault("venue_ids", {})[name] = venue_info
                else:
                    print(f"     ❌ Not found on Resy")
                    state.setdefault("venue_ids", {})[name] = None
                save_state(state)

            venue_info = state["venue_ids"].get(name)
            if not venue_info or not venue_info.get("venue_id"):
                print(f"  ⚠️  Skipping {name} — not found on Resy\n")
                continue

            venue_id = venue_info["venue_id"]
            print(f"\n  📍 {name} ({restaurant['area']}) — Resy ID: {venue_id}")

            for date in DATES:
                date_label = datetime.strptime(date, "%Y-%m-%d").strftime("%a %d %b")
                sys.stdout.write(f"     {date_label}... ")
                sys.stdout.flush()

                slots = get_resy_availability(venue_id, date)

                if slots:
                    key = f"{name}|resy|{date}"
                    existing = state.get("found", {}).get(key, [])
                    new_slots = [s for s in slots if s not in existing]

                    if new_slots:
                        new_findings.append({
                            "restaurant": name,
                            "area": restaurant["area"],
                            "platform": "resy",
                            "date": date,
                            "date_label": date_label,
                            "slots": slots,
                        })
                    state.setdefault("found", {})[key] = slots
                    print(f"✅ {len(slots)} slot(s): {', '.join(slots)}")
                else:
                    print("❌ None")

        elif platform == "opentable":
            print(f"\n  📍 {name} ({restaurant['area']}) — OpenTable (limited)")
            for date in DATES:
                date_label = datetime.strptime(date, "%Y-%m-%d").strftime("%a %d %b")
                sys.stdout.write(f"     {date_label}... ")
                sys.stdout.flush()
                slots = check_opentable_availability(restaurant, date)
                print(f"⚠️  OpenTable requires browser (manual check needed)")
                break  # Only need to say this once

    state["lastRun"] = datetime.utcnow().isoformat()
    save_state(state)

    print()
    if new_findings:
        print("🎉 NEW AVAILABILITY FOUND:\n")
        notify_parts = []
        for f in new_findings:
            print(f"  🍽️  {f['restaurant']} ({f['area']}) — {f['date_label']}")
            print(f"     Times: {', '.join(f['slots'])}")
            print()
            notify_parts.append(f"{f['restaurant']} on {f['date_label']} at {'/'.join(f['slots'][:3])}")
        print(f"NOTIFY:{'; '.join(notify_parts)}")
    else:
        print("No new availability found.")
        print("NOTIFY:none")


if __name__ == "__main__":
    main()
